### ISSUE TITLE
GIF frames: background, transparency, combining

### ISSUE THREAD

**Issue by Dragorn421** [python-pillow/Pillow#4640]
* OS: had the issue on Debian, moved to Windows for easier testing (I would say OS doesn't matter here)
* Python: 3.7.3 on Debian, 3.6.9 on Windows
* Pillow: 7.1.2

This was originally [a question on StackOverflow](https://stackoverflow.com/questions/61958291/gifs-read-with-pillow-have-black-borders), but I since think it's more of a Pillow bug/limitation, I will put all relevant information here, no need to read the SO post.

Resources:
[PepePls.gif](https://cdn.betterttv.net/emote/55898e122612142e6aaa935b/3x) 
![PepePls](https://user-images.githubusercontent.com/5306619/82728638-efa76d80-9cf1-11ea-89bf-84356d91743f.gif)

[PeepoCreepo.gif](https://cdn.betterttv.net/emote/5b369beff037d44c5936ff13/3x) 
![PeepoCreepo](https://user-images.githubusercontent.com/5306619/82728645-0352d400-9cf2-11ea-9b6f-34ac986ec1c7.gif)

```python
from PIL import Image

image = Image.open('PepePls.gif')

frames = []
try:
    while True:
        frames.append(image.copy())
        image.seek(image.tell() + 1)
except EOFError:
    pass

for i in range(len(frames)):
    frames[i].save(f'gif_f{i}.gif')
    frames[i].save(f'png_f{i}.png')
    frames[i].convert(mode='RGBA').save(f'rgba_gif_f{i}.gif')
    frames[i].convert(mode='RGBA').save(f'rgba_png_f{i}.png')
```

With PepePls.gif:
![image](https://user-images.githubusercontent.com/5306619/82728725-9c81ea80-9cf2-11ea-845d-8f507d479a16.png)

With PeepoCreepo.gif:
![peepocreepo_frames](https://user-images.githubusercontent.com/5306619/82728849-5f6a2800-9cf3-11ea-97b0-1712b861103d.png)

(I would shorten the gifs if I knew how to recreate their features, but I don't)

# Black borders
The first thing you notice is that there are black borders everywhere. After researching and reading through [GifImagePlugin.py](https://github.com/python-pillow/Pillow/blob/master/src/PIL/GifImagePlugin.py), I think this is because in `disposal_method==2` (which I think is "replace" whole frame) the load-image code (which I didn't locate precisely, I guess it's C somewhere?) pastes to a new image only the `dispose_extent` part of a frame which is the part read from the file (aka the part in the middle of the black borders). So that's fine but **I guess the "new image" which is being pasted on is likely initialized wrongly**, since black seems to be the default color in Pillow it may not be initialized at all.

I fixed this by copying the `dispose_extent` part onto a fully transparent image, but I'm not sure how to detected if the gif should be read that way (**how to tell if it wants a transparent background?**), code:
```python
from PIL import Image

image = Image.open('PeepoCreepo.gif')

frames = []
try:
	while True:
		if image.__dict__.get('disposal_method', 0) == 2:
			frame = Image.new('RGBA', image.size, color=(0,0,0,0))
			frame.paste(image.crop(image.dispose_extent), box=(image.dispose_extent[0],image.dispose_extent[1]))
			frames.append(frame)
		else:
			frames.append(image.copy())
		image.seek(image.tell() + 1)
except EOFError:
	pass

for i in range(len(frames)):
	frames[i].save(f'gif_f{i}.gif')
	frames[i].save(f'png_f{i}.png')
	frames[i].convert(mode='RGBA').save(f'rgba_gif_f{i}.gif')
	frames[i].convert(mode='RGBA').save(f'rgba_png_f{i}.png')
```

![image](https://user-images.githubusercontent.com/5306619/82729248-4f077c80-9cf6-11ea-9c7a-c6ec772f4291.png)

![peepocreepo_frames_dispose2](https://user-images.githubusercontent.com/5306619/82729324-ed93dd80-9cf6-11ea-8de9-ccbef7b4800f.png)

Apart from png_f2.png from PeepoCreepo.gif (no idea what is happening there either), all the black borders have disappeared when saving to png.

# gif and rgba
I'm not sure why `.convert(rgba).save(gif)` makes the saved image file lose transparency. Didn't research that issue at all.

# frame combine
Let's zoom in on the rgba_png_fX.png files, because I'm sure they went unnoticed among the hundreds of frames I dumped:
![image](https://user-images.githubusercontent.com/5306619/82729405-9e9a7800-9cf7-11ea-93a2-ff3606158a9c.png)
Notice rgba_png_f2.png has missing pixels
Here's a screenshot of GIMP (image manipulation program) with those frames which may help understand the issue:
![image](https://user-images.githubusercontent.com/5306619/82729459-00f37880-9cf8-11ea-8fbc-aa87fe8f2468.png)
Notice how in GIMP frame "Image vidéo 3" (which corresponds to our rgba_png_f2.png file) is "combine"
However if you add `print(image.__dict__.get('disposal_method', None))` at the beginning of the `while True:` loop you'll get
```
1
1
2
[... just 2]
2
1
```
I assume [1 is "combine"](https://github.com/python-pillow/Pillow/blob/master/src/PIL/GifImagePlugin.py#L301) ([at time of writing](https://github.com/python-pillow/Pillow/blob/dda6145fce8fbbe82db4d92cde41a64136ec02df/src/PIL/GifImagePlugin.py#L301)) and [2 is "replace"](https://github.com/python-pillow/Pillow/blob/master/src/PIL/GifImagePlugin.py#L264) ([at time of writing](https://github.com/python-pillow/Pillow/blob/dda6145fce8fbbe82db4d92cde41a64136ec02df/src/PIL/GifImagePlugin.py#L264))
So, **Pillow is using the `disposal_method` value early?** If it was delayed by a frame, rgba_png_f2.png would be correctly overlayed on rgba_png_f1.png

### my full "debugging" script (can be ignored)
For reference, here's a script with a lot of useless stuff which I used for debugging, it does highlight what undocumented `GifImageFile` members seem to be relevant here though. It also fixes the "delay combine" issue with a local variable disposal_method_last

```python
from PIL import Image

gif = 'PeepoCreepo.gif'
#gif = 'PepePls.gif'

image = Image.open(gif)

red = Image.new('RGBA', (85, 112), color='red')
redP = red.convert(mode='P')

imTest = Image.new('P', (10, 10), color='red')
imTest.im = Image.core.fill('P', imTest.size, 1)
#imTest.save('imtest.png')

# is this correct?
def getPaletteColor(im, n):
	# https://github.com/python-pillow/Pillow/blob/master/src/PIL/ImagePalette.py#L108
	try:
		return (im.palette.palette[n], im.palette.palette[n+256], im.palette.palette[n+512])
	except IndexError:
		return f'IndexError palette {n}'

loop = []
frames = []
durations = []

disposal_method_last = 0

try:
	while True:
		"""
		print('im=' + str(image.__dict__.get('im', None)))
		tile = image.__dict__.get('tile', None)
		print('tile=' + str(tile))
		print('tile[0]=' + str(tile[0]))
		# probably better than image.dispose_extent but idk
		print('tile[0][1]=' + str(tile[0][1]))
		"""
		print('dispose=' + str(image.__dict__.get('dispose', None)))
		print('dispose_extent=' + str(image.__dict__.get('dispose_extent', None)))
		print('disposal_method=' + str(image.__dict__.get('disposal_method', None)))
		"""
		print('background=' + str(image.info.get('background', None)))
		print('background color=' + str(getPaletteColor(image, image.info.get('background', 0))))
		print('transparency=' + str(image.info.get('transparency', None)))
		print('transparency color=' + str(getPaletteColor(image, image.info.get('transparency', 0))))
		# indeed the same as transparency
		print('upleft pixel dispose=' + str(image.getpixel((image.dispose_extent[0],image.dispose_extent[1]))))
		print('upleft_xy+1 pixel dispose=' + str(image.getpixel((image.dispose_extent[0]+1,image.dispose_extent[1]+1))))
		# indeed not transparency
		print('upleft pixel=' + str(image.getpixel((0,0))))
		print('upleft_xy+1 pixel=' + str(image.getpixel((1,1))))
		"""
		loop.append(image.info.get('loop', None))
		disposal_method = disposal_method_last
		disposal_method_last = image.__dict__.get('disposal_method', 0)
		if disposal_method == 2 or (disposal_method == 1 and frames == []):
			#"""
			frame = Image.new('RGBA', image.size, color=(255,0,0,0))
			"""
			frame = Image.new('P', image.size, color=image.info['background'])
			frame.palette = image.palette
			#"""
			frame.paste(image.crop(image.dispose_extent), box=(image.dispose_extent[0],image.dispose_extent[1]))
		elif disposal_method == 1:
			"""
			newStuff = Image.new('RGBA', image.size, color=(0,0,255,0))
			newStuff.paste(image.crop(image.dispose_extent), box=(image.dispose_extent[0],image.dispose_extent[1]))
			frame = Image.alpha_composite(frames[-1].convert(mode='RGBA'), newStuff.convert(mode='RGBA'))
			"""
			newStuff = image.crop(image.dispose_extent)
			frame = frames[-1].copy()
			"""
			background = Image.new('P', newStuff.size, color=image.info['background'])
			background.palette = image.palette
			frame.paste(background, image.dispose_extent)
			#"""
			frame.paste(newStuff, image.dispose_extent, newStuff.convert("RGBA"))
			#"""
		else:
			frame = image.copy()
		transparency = image.info.get('transparency', None)
		
		# "delta frames" not for PepePls apparently, or not how they work
		"""
		if len(frames) == 0:
			frame = image.convert(mode='RGBA')
		else:
			frame = Image.alpha_composite(frames[-1], image.convert(mode='RGBA'))
		"""
		# https://stackoverflow.com/questions/4904940/python-converting-gif-frames-to-png
		# didn't help
		"""
		if palette is None:
			palette = frame.getpalette()
		else:
			frame.putpalette(palette)
		"""
		frames.append(frame)
		durations.append(image.info.get('duration', None))
		"""
		"background" -> no difference
		"transparency" -> most borders changed color
		image.im = Image.core.fill("P", image.size, image.info["transparency"])
		image._prev_im = Image.core.fill("P", image.size, image.info["transparency"])
		"""
		"""
		same as above
		image.im = Image.new('P', image.size, color=200).im
		image._prev_im = Image.new('P', image.size, color=200).im
		"""
		image.seek(image.tell() + 1)
except EOFError:
	pass

for i in range(len(frames)):
	print(f'#{i} loop = {loop[i]} duration = {durations[i]} mode = {frames[i].mode} size = {frames[i].size}')
	frames[i].save(f'gif_f{i}.gif')
	frames[i].save(f'png_f{i}.png')
	frames[i].convert(mode='RGBA').save(f'rgba_gif_f{i}.gif')
	frames[i].convert(mode='RGBA').save(f'rgba_png_f{i}.png')
```

**Comment by karolyi** [comment]
@Dragorn421 you might want to look into #4644

**Comment by Dragorn421** [comment]
My issue is about loading gifs, not saving them, although in those tests I did you can indeed see gifs in Pillow hate transparency. Still, thanks

**Comment by karolyi** [comment]
My experience is if you convert a 'P' image to an RGBA, you get proper alpha values, except for when not (see also the issue for it).

Did you try your gifs with my processor? Also, in my experience, only the `disposal=2` works, none other does, unfortunately.

**Comment by Dragorn421** [comment]
I just did even though there's no point, it has the same issues as Pillow (black borders and bad frame combine), since it doesn't do anything different for loading gifs. It also looses information on frame duration (makes them all the same).

![outpeepocreepo](https://user-images.githubusercontent.com/5306619/82766881-0f8a7e80-9e23-11ea-9c09-3da58e670ffb.gif)

![outpepepls](https://user-images.githubusercontent.com/5306619/82766885-13b69c00-9e23-11ea-9843-892642e786da.gif)

**Comment by StupidCrow-yuan** [comment]
![gif_f74](https://user-images.githubusercontent.com/30788724/126296071-5f48136a-47be-44dc-8af0-3566b64877fa.gif)

I still get the error result! do you have any suggestion for me?

**Comment by Dragorn421** [comment]
Until a proper fix I guess you can use what I posted on stackoverflow https://stackoverflow.com/questions/61958291/gifs-read-with-pillow-have-black-borders under "EDIT 2:"
It was a long time ago so I don't have more to say than what I wrote here and over on Stackoverflow

**Comment by StupidCrow-yuan** [comment]
> Until a proper fix I guess you can use what I posted on stackoverflow https://stackoverflow.com/questions/61958291/gifs-read-with-pillow-have-black-borders under "EDIT 2:"
> It was a long time ago so I don't have more to say than what I wrote here and over on Stackoverflow

yeah, I have tested your "EDIT 2" on stackouverflow and test your "EDIT 3"  on github, but I still get other background color

**Comment by AnasYasin** [comment]
Any solutions guys? I'm getting the same error results on large gifs with more colors. After the first frame all frames are getting black background.   

**Code**:

```
gif = Image.open("gif4.GIF")
duration = 100
disposal = 2

frames = []
ind = 0
for frame in ImageSequence.Iterator(gif):
    bg2 = Image.open('test1.png')
    bg2.paste(frame, frame.convert('RGBA'))
    bg2.save('{}.png'.format(ind))
    frames.append(bg2)
    ind  += 1
    
frames[0].save('test.gif', save_all=True, append_images=frames[1:], loop=0, optimise=False, duration=duration, disposal=disposal, format='GIF')
```

**Orignal Gif**:
<img src="https://user-images.githubusercontent.com/34492965/136784497-c7c61318-50fa-47f3-b943-5a9dd18193db.GIF" width=20% height=20%>

**Background Image**:
<img src="https://user-images.githubusercontent.com/34492965/136784783-b8b53bff-f79e-417d-9b74-3dbd25f0e8c1.png" width=20% height=20%>

**Final Result**:
<img src="https://user-images.githubusercontent.com/34492965/136785161-d72f6fb6-a535-4eac-a02f-96b58bdafce2.gif" width=20% height=20%>

**Frames**:

<img src="https://user-images.githubusercontent.com/34492965/136785366-3588abdf-d702-4e7d-8452-4ceb2fb7bc22.png" width=10% height=10%> <img src="https://user-images.githubusercontent.com/34492965/136785383-c5977161-3e8d-4dfd-aaa8-2dfd9865215b.png" width=10% height=10%> <img src="https://user-images.githubusercontent.com/34492965/136785389-f4f61c4a-bc4e-444c-a7a2-96e8ef6babbb.png" width=10% height=10%> <img src="https://user-images.githubusercontent.com/34492965/136785393-cc2137d5-3256-45a3-b53e-22769cad1d9b.png" width=10% height=10%> 
 <img src="https://user-images.githubusercontent.com/34492965/136785396-ada9d6b9-4c63-4614-bc39-eb9254ce9bf8.png" width=10% height=10%>

**Comment by tomhaydn** [comment]
I have the exact same issue as above:

`from PIL import Image, ImageSequence, GifImagePlugin
import os


bg = Image.open('images/bg/Background Sky.gif')
body = Image.open('images/body/Body Blue.gif')
eyes = Image.open('images/eyes/Eyes Green.gif')
headwear = Image.open('images/headwear/Headwear Party Hat.gif')

frames = []
for idx, frame in enumerate(ImageSequence.Iterator(bg)):
    frame = Image.new('RGBA', (640, 640), color=(255,0,0,0))
    frame.paste(ImageSequence.Iterator(bg)[idx])
    frame.paste(ImageSequence.Iterator(body)[idx], mask=ImageSequence.Iterator(body)[idx].convert("RGBA"))
    frame.paste(ImageSequence.Iterator(eyes)[idx], mask=ImageSequence.Iterator(eyes)[idx].convert("RGBA"))
    frame.paste(ImageSequence.Iterator(headwear)[idx], mask=ImageSequence.Iterator(headwear)[idx].convert("RGBA"))
    
    frames.append(frame)

print(frames)

#frames[0].save('images/generated/output.gif', save_all=True, append_images=frames[0:])
frames[0].save('images/generated/output.gif', save_all=True, append_images=frames[1:], loop=0)`

**Comment by radarhere** [comment]
The issue from @tomhaydn became #5837 and the issue from @AnasYasin became #5755

**Comment by radarhere** [comment]
With #5857, the PNGs look correct.

**Comment by radarhere** [comment]
#5859 fixes the background color in the GIFs saved from the first image after converting to RGBA.

**Comment by radarhere** [comment]
#5857 has now been merged.

With #5859 and #5869, this should work correctly.

**Comment by radarhere** [comment]
Those PRs has been merged, so this should be fixed in Pillow 9.0, due to be released on January 2.
